Begun work on support for running in a container.

QuadmanSWE opened the pull request:
- Added Dockerfile
- Wrote build and usage instructions for kube-diagrams and helm-diagrams from a manifest and a remote helm chart respectively.

QuadmanSWE commented:
@philippemerle thanks for the request to write this.

This is taking KubeDiagrams directly through pip, which might not be exactly what you want.

I am not sure exactly how you publish to pip, but I would recommend that you build and publish to dockerhub in the same pipeline if all your tests pass. Perhaps add some tests for executing in the container to make sure helm and all the dependencies work together as expected.

Just let me know if you have any need of a second pair of eyes when working with your CI/CD pipeline :)

QuadmanSWE commented:
The future goal for me with having this great visulization is to run it on demand, perhaps with knative or something, as a rendering service for parts of the stuff running in a cluster. Sort of like how argocd renders its applications, but for a internal developer portal to display resources that are deployed. The idea is to query kubeapi by label from the perspective of the component/system/team that is being viewed in the portal.

philippemerle commented:
@QuadmanSWE: Thank you for this excellent contribution!

> This is taking KubeDiagrams directly through pip, which might not be exactly what you want.

Right for now! Later we will updated it to copy the `kube-diagrams` and `helm-diagrams` scripts directly from the repo.

> I am not sure exactly how you publish to pip, 

Currently, I publish it manually. Later this will be done automitically via a CI.

> but I would recommend that you build and publish to dockerhub in the same pipeline if all your tests pass. Perhaps add some tests for executing in the container to make sure helm and all the dependencies work together as expected.

I will take these into account.

> Just let me know if you have any need of a second pair of eyes when working with your CI/CD pipeline :)

Thank you for your help.

> The future goal for me with having this great visulization is to run it on demand, perhaps with knative or something, as a rendering service for parts of the stuff running in a cluster. Sort of like how argocd renders its applications, but for a internal developer portal to display resources that are deployed. The idea is to query kubeapi by label from the perspective of the component/system/team that is being viewed in the portal.

Very interesting future goal, which is a new usage for KubeDiagrams.

philippemerle commented:
Hi @QuadmanSWE,
I tried to build the image with `docker` instead of `podman`:
```ssh
$ docker build -t kd .
...
 > [base 2/7] RUN apt-get update && apt-get -y install graphviz libgl1:                                                                                                                  
#6 0.294 Get:1 http://deb.debian.org/debian bookworm InRelease [151 kB]                                                                                                                  
#6 0.351 Get:2 http://deb.debian.org/debian bookworm-updates InRelease [55.4 kB]                                                                                                         
#6 0.372 Get:3 http://deb.debian.org/debian-security bookworm-security InRelease [48.0 kB]                                                                                               
#6 0.423 Err:1 http://deb.debian.org/debian bookworm InRelease                                                                                                                           
#6 0.423   The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 0E98404D386FA1D9 NO_PUBKEY 6ED0E7B82643E131 NO_PUBKEY F8D2585B8783D481
#6 0.482 Err:2 http://deb.debian.org/debian bookworm-updates InRelease
#6 0.482   The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 0E98404D386FA1D9 NO_PUBKEY 6ED0E7B82643E131
#6 0.540 Err:3 http://deb.debian.org/debian-security bookworm-security InRelease
#6 0.540   The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 54404762BBB6E853 NO_PUBKEY BDE6D2B9216EC7A8
#6 0.542 Reading package lists...
#6 0.553 W: GPG error: http://deb.debian.org/debian bookworm InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 0E98404D386FA1D9 NO_PUBKEY 6ED0E7B82643E131 NO_PUBKEY F8D2585B8783D481
#6 0.553 E: The repository 'http://deb.debian.org/debian bookworm InRelease' is not signed.
#6 0.553 W: GPG error: http://deb.debian.org/debian bookworm-updates InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 0E98404D386FA1D9 NO_PUBKEY 6ED0E7B82643E131
#6 0.553 E: The repository 'http://deb.debian.org/debian bookworm-updates InRelease' is not signed.
#6 0.553 W: GPG error: http://deb.debian.org/debian-security bookworm-security InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 54404762BBB6E853 NO_PUBKEY BDE6D2B9216EC7A8
#6 0.553 E: The repository 'http://deb.debian.org/debian-security bookworm-security InRelease' is not signed.
#6 0.553 E: Problem executing scripts APT::Update::Post-Invoke 'rm -f /var/cache/apt/archives/*.deb /var/cache/apt/archives/partial/*.deb /var/cache/apt/*.bin || true'
#6 0.553 E: Sub-process returned an error code
------
failed to solve with frontend dockerfile.v0: failed to build LLB: executor failed running [/bin/sh -c apt-get update && apt-get -y install graphviz libgl1]: runc did not terminate sucessfully
```
It seems that public keys are missed. Have you any idea about this issue?

I replaced `FROM docker.io/python:3.12-slim AS base` by `FROM docker.io/python:3.12-bullseye AS base`in `Dockerfile`, then I successed to build the image but the image is very fat (i.e. 1.56GB). I successed to build an image from `docker.io/python:3.12-alpine` and then the image is smaller (i.e. 240MB).

Is the `libgl1` apt package really required? Are the `helm-python` `glib` `PyQt5` `graphviz` Python packages really required?

QuadmanSWE commented:
No idea. Like I wrote before I just kept adding packages until I got it to
work.

I initially wanted to create a multi stage build and just have the binaries
and dependencies copied over which is why I started with slim.

I ended up with so many issues though due to libraries missing and since I
know so little about python dependencies I just ended up with the one
image. I never had the issue you are describing so I never jumped to any
other release.

I am happy that it worked on alpine. :)

Den lör 15 mars 2025 08:36Philippe Merle ***@***.***> skrev:

> Hi @QuadmanSWE <https://github.com/QuadmanSWE>,
> I tried to build the image with docker instead of podman:
>
> $ docker build -t kd .
> ...
>  > [base 2/7] RUN apt-get update && apt-get -y install graphviz libgl1:
> #6 0.294 Get:1 http://deb.debian.org/debian bookworm InRelease [151 kB]
> #6 0.351 Get:2 http://deb.debian.org/debian bookworm-updates InRelease [55.4 kB]
> #6 0.372 Get:3 http://deb.debian.org/debian-security bookworm-security InRelease [48.0 kB]
> #6 0.423 Err:1 http://deb.debian.org/debian bookworm InRelease
> #6 0.423   The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 0E98404D386FA1D9 NO_PUBKEY 6ED0E7B82643E131 NO_PUBKEY F8D2585B8783D481
> #6 0.482 Err:2 http://deb.debian.org/debian bookworm-updates InRelease
> #6 0.482   The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 0E98404D386FA1D9 NO_PUBKEY 6ED0E7B82643E131
> #6 0.540 Err:3 http://deb.debian.org/debian-security bookworm-security InRelease
> #6 0.540   The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 54404762BBB6E853 NO_PUBKEY BDE6D2B9216EC7A8
> #6 0.542 Reading package lists...
> #6 0.553 W: GPG error: http://deb.debian.org/debian bookworm InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 0E98404D386FA1D9 NO_PUBKEY 6ED0E7B82643E131 NO_PUBKEY F8D2585B8783D481
> #6 0.553 E: The repository 'http://deb.debian.org/debian bookworm InRelease' is not signed.
> #6 0.553 W: GPG error: http://deb.debian.org/debian bookworm-updates InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 0E98404D386FA1D9 NO_PUBKEY 6ED0E7B82643E131
> #6 0.553 E: The repository 'http://deb.debian.org/debian bookworm-updates InRelease' is not signed.
> #6 0.553 W: GPG error: http://deb.debian.org/debian-security bookworm-security InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 54404762BBB6E853 NO_PUBKEY BDE6D2B9216EC7A8
> #6 0.553 E: The repository 'http://deb.debian.org/debian-security bookworm-security InRelease' is not signed.
> #6 0.553 E: Problem executing scripts APT::Update::Post-Invoke 'rm -f /var/cache/apt/archives/*.deb /var/cache/apt/archives/partial/*.deb /var/cache/apt/*.bin || true'
> #6 0.553 E: Sub-process returned an error code
> ------
> failed to solve with frontend dockerfile.v0: failed to build LLB: executor failed running [/bin/sh -c apt-get update && apt-get -y install graphviz libgl1]: runc did not terminate sucessfully
>
> It seems that public keys are missed. Have you any idea about this issue?
>
> I replaced FROM docker.io/python:3.12-slim AS base by FROM
> docker.io/python:3.12-bullseye AS basein Dockerfile, then I successed to
> build the image but the image is very fat (i.e. 1.56GB). I successed to
> build an image from docker.io/python:3.12-alpine and then the image is
> smaller (i.e. 240MB).
>
> Is the libgl1 apt package really required? Are the helm-python glib PyQt5
> graphviz Python packages really required?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/philippemerle/KubeDiagrams/pull/15#issuecomment-2726309271>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB5SHXNOVHDCGJP3HKTNW2D2UPJ7VAVCNFSM6AAAAABZAO7252VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMRWGMYDSMRXGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: philippemerle]*philippemerle* left a comment
> (philippemerle/KubeDiagrams#15)
> <https://github.com/philippemerle/KubeDiagrams/pull/15#issuecomment-2726309271>
>
> Hi @QuadmanSWE <https://github.com/QuadmanSWE>,
> I tried to build the image with docker instead of podman:
>
> $ docker build -t kd .
> ...
>  > [base 2/7] RUN apt-get update && apt-get -y install graphviz libgl1:
> #6 0.294 Get:1 http://deb.debian.org/debian bookworm InRelease [151 kB]
> #6 0.351 Get:2 http://deb.debian.org/debian bookworm-updates InRelease [55.4 kB]
> #6 0.372 Get:3 http://deb.debian.org/debian-security bookworm-security InRelease [48.0 kB]
> #6 0.423 Err:1 http://deb.debian.org/debian bookworm InRelease
> #6 0.423   The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 0E98404D386FA1D9 NO_PUBKEY 6ED0E7B82643E131 NO_PUBKEY F8D2585B8783D481
> #6 0.482 Err:2 http://deb.debian.org/debian bookworm-updates InRelease
> #6 0.482   The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 0E98404D386FA1D9 NO_PUBKEY 6ED0E7B82643E131
> #6 0.540 Err:3 http://deb.debian.org/debian-security bookworm-security InRelease
> #6 0.540   The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 54404762BBB6E853 NO_PUBKEY BDE6D2B9216EC7A8
> #6 0.542 Reading package lists...
> #6 0.553 W: GPG error: http://deb.debian.org/debian bookworm InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 0E98404D386FA1D9 NO_PUBKEY 6ED0E7B82643E131 NO_PUBKEY F8D2585B8783D481
> #6 0.553 E: The repository 'http://deb.debian.org/debian bookworm InRelease' is not signed.
> #6 0.553 W: GPG error: http://deb.debian.org/debian bookworm-updates InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 0E98404D386FA1D9 NO_PUBKEY 6ED0E7B82643E131
> #6 0.553 E: The repository 'http://deb.debian.org/debian bookworm-updates InRelease' is not signed.
> #6 0.553 W: GPG error: http://deb.debian.org/debian-security bookworm-security InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 54404762BBB6E853 NO_PUBKEY BDE6D2B9216EC7A8
> #6 0.553 E: The repository 'http://deb.debian.org/debian-security bookworm-security InRelease' is not signed.
> #6 0.553 E: Problem executing scripts APT::Update::Post-Invoke 'rm -f /var/cache/apt/archives/*.deb /var/cache/apt/archives/partial/*.deb /var/cache/apt/*.bin || true'
> #6 0.553 E: Sub-process returned an error code
> ------
> failed to solve with frontend dockerfile.v0: failed to build LLB: executor failed running [/bin/sh -c apt-get update && apt-get -y install graphviz libgl1]: runc did not terminate sucessfully
>
> It seems that public keys are missed. Have you any idea about this issue?
>
> I replaced FROM docker.io/python:3.12-slim AS base by FROM
> docker.io/python:3.12-bullseye AS basein Dockerfile, then I successed to
> build the image but the image is very fat (i.e. 1.56GB). I successed to
> build an image from docker.io/python:3.12-alpine and then the image is
> smaller (i.e. 240MB).
>
> Is the libgl1 apt package really required? Are the helm-python glib PyQt5
> graphviz Python packages really required?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/philippemerle/KubeDiagrams/pull/15#issuecomment-2726309271>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB5SHXNOVHDCGJP3HKTNW2D2UPJ7VAVCNFSM6AAAAABZAO7252VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMRWGMYDSMRXGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

philippemerle commented:
Following is my modified `Dockerfile`
```
FROM docker.io/alpine/helm:3 AS helm

FROM docker.io/python:3.12-alpine AS base
RUN apk update && apk add graphviz
RUN pip install --upgrade pip && pip install PyYAML diagrams
ADD bin /usr/local/bin/
COPY --from=helm /usr/bin/helm /usr/local/bin/helm
WORKDIR /work
```
This is based on Alpine to have a small image, this installs the minimum of dependencies, and uses the latest KubeDiagrams  scripts to benefit of last improvements.

Following illustrates the new usage:
```ssh
$ docker run \
      -v "$PWD":/work kd \
      kube-diagrams -o cassandra.png examples/cassandra/cassandra.yml
```
As result, the generated diagram is automatically save into the host file system, i.e., don't need to name the container and copy the generated diagram.

From a Helm chart:
```ssh
$ docker run \
      -v "$PWD":/work kd \
      helm-diagrams https://charts.jetstack.io/cert-manager
```
The generated diagram is directly available into the host file system.

I will improve the `helm-diagrams` script to avoid the requirement of the `KubeDiagrams.yaml` file, i.e., used only if this file is present.

What do you think about this version? Is it matching your needs?

QuadmanSWE commented:
Looks good. I will not have the time to try it out until monday.

I will keep you updated on my project once I have made some more progress.

philippemerle commented:
@QuadmanSWE: official docker images for KubeDiagrams wiil be available in https://hub.docker.com/r/philippemerle/kubediagrams. I updated Dockerfile and document usage in README.md. I will add a github action to  build the latest image and push it at each github push automatically.

QuadmanSWE commented:
@philippemerle tried it out now, your four examples all worked well for me. **You can remove the podman part, it doesn't add anything and the way I tested it is not a good way to run it anyway**. I really liked your example of pipeing the yaml from a cluster into the input of the execution in the container. That is exactly what I needed. 
`# generate a diagram from the actual cluster state
kubectl get all -n demo -l app=demo-backstage -o yaml | podman run -v "$(pwd)":/work -i docker.io/philippemerle/kubediagrams kube-diagrams -o demo-backstage.png -
`
![image](https://github.com/user-attachments/assets/818ddf22-54f5-4d91-ba8e-1bb71e4b5519)

What's the procedure to get additional CRDs into this tool? for example [CNPG clusters](https://cloudnative-pg.io/)?
`Warning: No workload resource matches labels({'cnpg.io/cluster': 'keycloak', 'role': 'replica'}) for Service(keycloak-ro)!`

philippemerle commented:
> @philippemerle tried it out now, your four examples all worked well for me.

Good news.

> You can remove the podman part, it doesn't add anything and the way I tested it is not a good way to run it anyway.

Ok I will do.

> I really liked your example of pipeing the yaml from a cluster into the input of the execution in the container. That is exactly what I needed.

Thank to @himslm01 who contributed this feature.

Note that red edges and red labelled nodes will be deleted soon, as they pollute diagrams, missing nodes will be only reported in the console.

> What's the procedure to get additional CRDs into this tool? for example [CNPG clusters](https://cloudnative-pg.io/)?

You should create a KubeDiagrams custom config file as follows:
```yaml
nodes:
    Cluster/postgresql.cnpg.io/v1:
        scope: Namespaced # or Cluster
        diagram_node_classname: diagrams.onprem.database.Postgresql
        # or custom_icon: filename-of-the-custom-icon.png
        nodes: \
            # Python code to create children nodes, see an example in bin/kube-diagrams.yaml
       edges: \
           # Python code to create edges to other nodes, see examples in bin/kube-diagrams.yaml
    # other node declarations
```

Note that Python code in `nodes` and `edges` will be modified in the future, see discussion https://github.com/philippemerle/KubeDiagrams/discussions/6.

To use this custom config file:
```ssh
kube-diagrams -c the-custom-config-filename .... other CLI parameters ...
```

> Warning: No workload resource matches labels({'cnpg.io/cluster': 'keycloak', 'role': 'replica'}) for Service(keycloak-ro)!

This warning is raised because no workload resource matches the podSelector of the Service object named keycloak-ro. If your manifests contain some workload resources matching this selector then the selector algo in KubeDiagrams is bugged or needed to be improved (please report an issue with an example).
